### ISSUE TITLE
Policy gradients do not assume single-step decisions; terminal reward can train every latent recurrence

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#4]
Reviewed commit 2409067b59ff4b0e6a7833a3c79c2299355a5335.

Affected: [map.html, credit-assignment gap](https://github.com/microprediction/latentreasoning/blob/2409067b59ff4b0e6a7833a3c79c2299355a5335/map.html#L80-L86), the rltt gap edge, and the Williams–Türeci bibliography annotation in index.html.

The map says policy-gradient objectives implicitly assume single-step decisions and therefore mis-specify multi-step latent computation. This is false as a mathematical statement.

For a differentiable deterministic recurrence
$$
z_{j+1}=F_\theta(z_j,x),\qquad y\sim\pi_\theta(\cdot\mid z_K),
$$
with reward independent of $\theta$ conditional on $y$,
$$
\nabla_\theta E[R(y)]
=E[R(y)\nabla_\theta\log\pi_\theta(y\mid z_K(\theta))].
$$
The total derivative includes every unrolled latent update when gradients are retained. A terminal objective does not imply a gradient only at the terminal state.

An elementary counterexample is $z_0=0,\ z_{j+1}=z_j+\theta$, followed by $y\sim\mathrm{Bernoulli}(\sigma(z_K))$, with reward $R=y$. Then
$$
J(\theta)=\sigma(K\theta),\qquad
J'(\theta)=K\sigma(K\theta)(1-\sigma(K\theta)).
$$
All K updates contribute despite a single terminal reward.

For stochastic latent transitions, likelihood-ratio terms for those transitions (or suitable pathwise derivatives) are needed. Detaching states or truncating backpropagation can remove credit in an implementation. Sparse rewards and long products of Jacobians can also make learning difficult. These are meaningful issues, distinct from invalidity of the policy-gradient objective.

**Missing foundational literature:** [Wierstra et al., Recurrent Policy Gradients (2009 preprint)](https://people.idsia.ch/~juergen/joa2009.pdf) explicitly trains recurrent policies with return-weighted backpropagation.

**Treatment of the cited new method:** [Williams & Türeci, RLTT](https://arxiv.org/html/2602.10520v1) reports benefits from per-loop auxiliary credit. Its own explanation in Section 5.1.1 says the terminal signal backpropagates through latent refinements. Report the empirical gains and optimization hypothesis, but do not promote them into a proof that ordinary policy gradients cannot train recurrence. Appendix A.8 also already contains a conditional theoretical analysis, making the map's unqualified absence-of-theory claim especially awkward.

Suggested replacement gap: which auxiliary objectives and truncation/halting choices improve gradient quality and compute efficiency for latent reasoning, under controlled comparisons?

**Comment by microprediction** [comment]
Verified and **accepted**. Fixed in dbc3e8c.

The mathematical point is correct and the map's claim was false as stated. For a differentiable recurrence with gradients retained, the total derivative of a terminal reward already includes every unrolled update — your `z_{j+1} = z_j + θ`, `y ~ Bernoulli(σ(z_K))` example makes it unambiguous: `J'(θ) = Kσ(Kθ)(1−σ(Kθ))`, with all K updates contributing under a single terminal reward.

I also checked the RLTT paper rather than relying on my earlier summary of it, and you're right on both counts: §5.1.1 describes the terminal signal backpropagating through the latent refinements, and Appendix A.8 contains a conditional theoretical analysis. So "no theory exists" was wrong on the paper's own terms.

The gap is now recast as a question rather than an absence:

> **Which auxiliary objective, and why.** *Revised.* This previously claimed that policy-gradient objectives assume single-step decisions and so cannot train latent recurrence. That is false […] What is true is narrower — standard objectives have *empirically* underperformed on looped models, and the causes on offer (long Jacobian products, sparse reward, truncation or detaching in implementations) are optimisation problems, not invalidity.

The bibliography annotation for Williams & Türeci now reports the empirical finding and the optimisation hypothesis without promoting either into a proof, and **Wierstra, Förster, Peters & Schmidhuber (2009), Recurrent Policy Gradients** is added as the foundational precedent — cited precisely because it long predates and contradicts the claim.

**Comment by microprediction** [comment]
Follow-up at 3b15c6f2163049ac1cc0200909deb1ea027a5de4: the map's prose now correctly distinguishes a valid terminal-reward gradient from optimization difficulties. One occurrence remains in the [RLTT node tooltip](https://github.com/microprediction/latentreasoning/blob/3b15c6f2163049ac1cc0200909deb1ea027a5de4/map.html#L245-L246): it attributes policy-gradient failure to rewarding only the final latent state.

Please bring the tooltip into agreement with the corrected prose. Suggested replacement: “RLTT reports improvements over GRPO on Ouro-2.6B-Thinking by adding objectives at intermediate loops. Terminal-reward gradients can already flow through the recurrence; the reported benefit concerns optimization and credit assignment in the tested setup.”

The distinction is substantive: a loss evaluated at the final state still has total derivative through earlier states, as in the counterexample above. The [paper's empirical comparison](https://arxiv.org/html/2602.10520v1) does not establish universal failure of terminal policy gradients. No new issue is needed for this remaining occurrence.

**Comment by microprediction** [comment]
New propagation at current default commit `7466f79ef53ce00d7cff1dd1537bfffc7fedf345`: the newly added training page reintroduces the same mathematical error, now in its taxonomy of latent-RL estimators.

Affected passages:

- [`training.html` lines 190–205](https://github.com/microprediction/latentreasoning/blob/7466f79ef53ce00d7cff1dd1537bfffc7fedf345/training.html#L190-L205) say Coconut “supplies none” of the rollout distribution, log-probability or gradient requirements, and conclude that there is “no distribution π, no log-probability, and no score function.”
- [lines 210–216](https://github.com/microprediction/latentreasoning/blob/7466f79ef53ce00d7cff1dd1537bfffc7fedf345/training.html#L210-L216) contrast full-recurrence pathwise differentiation with the methods below and say **all** of them instead inject noise so a score-function gradient “touches only one step's computation.”
- [lines 232–280](https://github.com/microprediction/latentreasoning/blob/7466f79ef53ce00d7cff1dd1537bfffc7fedf345/training.html#L232-L280) consequently put Dropout-GRPO into that one-step stochastic-action taxonomy.

Dropout-GRPO itself demonstrates why this is false. Let its replayed mask be `ξ ~ p(ξ)`, with `p` fixed independently of `θ`; let `h_T(θ,ξ)` be the deterministic masked latent recurrence; and let the answer be sampled from `π_θ(y | x,h_T)`. Then

`J(θ) = E_{ξ,y}[R(y)]`

has policy gradient

`∇J(θ) = E[R(y) ∇_θ log π_θ(y | x,ξ)]`.

There is no mask score term because `∇_θ log p(ξ)=0`. The answer-token log-probability is the tractable policy likelihood, and its total derivative includes `h_T(θ,ξ)` and therefore the whole latent recurrence.

This is not merely an alternative derivation. [Jung, arXiv:2606.10184v1, §3.1–3.3](https://arxiv.org/html/2606.10184#S3) explicitly:

1. defines `Π_θ(ξ,y|x)=p(ξ)π_θ(y|x,ξ)` (Eq. 3);
2. gives the recursive full-chain Jacobian `∂h_t/∂θ` (Eq. 4);
3. says this recursion is what `loss.backward()` computes with the mask held constant;
4. differentiates `log π_θ(y|x,ξ)`, not a parameterized log-density for the mask.

So Dropout-GRPO takes the pathwise/BPTT route through every latent step **and** a score-function gradient for the sampled answer tokens. Dropout supplies exogenous rollout diversity; it does not replace BPTT with an independent per-latent score term or make the gradient touch only one step.

The stronger statement that deterministic Coconut rollouts necessarily make group advantages zero also needs a decoding qualification. Coconut's latent path is deterministic, but its terminal answer decoder still defines a token distribution. Independently sampling answer tokens can produce different rewards and a valid terminal policy gradient even with the same latent path. Advantage collapses only if the answer generation is also deterministic (or all sampled answers receive the same reward). Jung's paper explicitly inserts the “under low-temperature decoding” condition when asserting identical answers (§3.1); the page drops it.

The genuine limitation is narrower: Coconut has no native **per-latent-transition** likelihood and no stochastic halting action, so terminal-token GRPO cannot assign likelihood-ratio terms directly to latent steps or learn the horizon. It can still optimize the deterministic recurrence by backpropagating the terminal answer likelihood through it.

Suggested correction:

- Replace “Coconut supplies none” with: “Coconut supplies a terminal answer-token policy but no stochastic per-latent policy; standard terminal GRPO is valid but requires gradients through the unrolled recurrence and cannot directly optimize latent-step likelihoods or a fixed horizon.”
- State that embedding/Gumbel/diffusion methods create parameterized latent-action densities, whereas Dropout-GRPO uses exogenous mask diversity and full-recurrence pathwise differentiation of the answer-token likelihood.
- Qualify zero group advantage by deterministic/low-temperature answer decoding or equal rewards.

This also makes the new page agree with its own [lines 454–460](https://github.com/microprediction/latentreasoning/blob/7466f79ef53ce00d7cff1dd1537bfffc7fedf345/training.html#L454-L460), which correctly say terminal reward backpropagates through the loops.

**Comment by microprediction** [comment]
A further mathematical qualification at current default commit `11b700ba0c9cc03b0e1a54db15fe06d15fad8e51` belongs with this issue.

The new training page accurately calls RLTT’s terms “per-loop” credit, but [lines 438–446](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L438-L446) still present the weighted intermediate-loop log-probabilities as a replacement policy-gradient objective without stating that they are auxiliary/off-policy scores for tokens sampled from a different distribution.

In the primary paper, the actual rollout policy samples each token only from the terminal distribution:

```
y ~ P_theta^(T)
```

([§3.1, Eq. 1](https://arxiv.org/html/2602.10520#S3.SS1)). RLTT then adds, for every earlier loop `t < T`,

```
omega_t * A(y) * grad log P_theta^(t)(y)
```

([§3.2, Eq. 4](https://arxiv.org/html/2602.10520#S3.SS2)). But `y` was not sampled from `P_theta^(t)`. No importance ratio converts the terminal sample to the intermediate distribution. Therefore the earlier-loop terms are not likelihood-ratio estimators of the expected reward under either the deployed terminal policy or the intermediate policy. They are reward-weighted auxiliary cross-entropy/semi-gradient terms.

A two-action counterexample makes the distinction exact. Let the terminal policy be `P_T=(0.9,0.1)`; let an intermediate policy be `P_1=(0.5,0.5)`, parameterized by a binary logit `phi=0`; and let the reward be the constant `R=1`. The true expected reward is identically one, so its gradient is zero. Yet

```
E_{y~P_T}[d/dphi log P_1(y)]
= 0.9*(1-0.5) + 0.1*(0-0.5)
= 0.4.
```

Thus the RLTT intermediate term can be nonzero when the true policy gradient is zero. Calling `P^(t)` a “proper next-token distribution” does not repair the sampling mismatch.

This does **not** undermine the reported empirical comparison. It sharpens the interpretation already adopted in this issue and in the bibliography: RLTT adds useful per-loop auxiliary supervision; it is not an unbiased decomposition of the terminal policy gradient across latent states.

Suggested small correction to the training page:

> RLTT adds a weighted auxiliary log-score at each loop for the token sampled from the terminal-loop policy. These intermediate terms provide denser reward-weighted supervision, but because the token was sampled from the terminal distribution rather than each intermediate distribution, they are not separate likelihood-ratio terms of the deployed policy.

That wording would also prevent “process signal” from being read as a formally valid per-loop policy-gradient factorization.

**Comment by microprediction** [comment]
A new propagation at current default commit `88d1db6b1c014f641f67ec4b84ec1f51e8f0c43b` belongs with this issue.

The [Switch bibliography entry](https://github.com/microprediction/latentreasoning/blob/88d1db6b1c014f641f67ec4b84ec1f51e8f0c43b/bibliography.html#L1135-L1138) says:

> The policy ratio is undefined inside a latent block, so this wraps the block in two ordinary tokens to make entry and exit legible to on-policy reinforcement learning while the interior stays continuous.

This needs the same per-latent-versus-terminal qualification established above. A deterministic hidden-state recurrence has no **separate per-latent action density**, but that does not make the rollout likelihood or its policy ratio undefined. If the latent states are deterministic functions of the preceding sampled text and parameters, then

[
pi_	heta(	aumid x)
=prod_{tinmathcal T_{
m text}}
pi_	heta(a_tmid a_{<t},z_t(	heta)),
]

so the old/new ratio is defined over the sampled text decisions. Its total derivative can pass through every deterministic latent update in (z_t(	heta)), exactly as in the terminal-gradient derivation in the issue body.

This is also how the Switch paper formulates the method. [§3.3](https://arxiv.org/html/2606.13106#S3.SS3) states that the latent injection is deterministic, the rollout likelihood factorizes over text positions only, and latent positions contribute no direct policy-density term. The `<swi>` and `</swi>` tokens provide learned stochastic entry and exit decisions and align on-policy rollouts with deployment; they do not create a likelihood ratio for the continuous interior or make an otherwise undefined trajectory ratio defined.

Suggested replacement:

> Hidden-state recurrence has no separate per-latent action density. Switch adds discrete entry and exit tokens so GRPO can learn whether and when to invoke a deterministic latent block while gradients from sampled text decisions backpropagate through it; latent positions contribute no likelihood-ratio term.

That preserves the paper's genuine contribution—trainable switching and stopping—without implying that deterministic recurrence is intrinsically incompatible with an on-policy trajectory ratio.

**Comment by microprediction** [comment]
The immediately adjacent SLPO annotation at the same current commit has a second occurrence of the same accepted distinction.

[`bibliography.html` lines 1140–1143](https://github.com/microprediction/latentreasoning/blob/88d1db6b1c014f641f67ec4b84ec1f51e8f0c43b/bibliography.html#L1140-L1143) says that because latent trajectories lack a tractable per-step likelihood and stopping interface, “outcome-reward reinforcement learning cannot elicit latent test-time scaling as it does for explicit chains.”

The first premise is correctly scoped to **per-latent-transition** likelihood, and the missing adaptive stopping action does prevent learning the horizon. The conclusion is too strong: neither fact prevents outcome-reward optimization of a fixed-horizon deterministic recurrence through sampled answer tokens and backpropagation.

The SLPO paper itself makes this narrower structure explicit. Its rollout score includes all three of:

1. a Gaussian **surrogate** score for latent transitions;
2. the ordinary answer-token likelihood;
3. a stopping-time likelihood

([§4.2, Eqs. 10–11](https://arxiv.org/html/2607.19691#S4.SS2)). It also states that its realized latent states are stop-gradient targets and that agreement of the surrogate update with the expected-reward gradient requires a separate score-alignment assumption ([Appendix A.5, Propositions 2–4](https://arxiv.org/html/2607.19691#A5)). Thus SLPO adds direct transition scoring and learnable stopping; it does not establish that outcome rewards were unusable before those additions.

Suggested replacement:

> Fixed-horizon latent recurrence has no native per-transition likelihood or stopping action, so ordinary terminal outcome RL cannot assign separate likelihood-ratio credit to latent steps or learn how long to think. SLPO adds a moment-matched Gaussian surrogate for transition scoring and a learned stopping head, alongside the existing answer-token likelihood; connecting the surrogate update to expected reward requires the paper’s score-alignment condition.

This would also keep the bibliography consistent with the corrected training-page treatment of SLPO in issue #71 and with the terminal-gradient distinction already established here.
